File: conformance/code/individual-go/internal/clients/vault_auth.go
package clients

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
	"time"
)

// vaultAuth manages a Vault token obtained via AppRole login and refreshes it
// on demand (see ensureToken) — there is no background goroutine. The service
// only ever performs encrypt/decrypt; all key lifecycle (rotation, config) is
// managed by Vault itself.
type vaultAuth struct {
	address  string
	roleID   string
	secretID string
	http     *http.Client

	mu        sync.RWMutex
	tok       string
	renewable bool
	ttl       time.Duration
	expiresAt time.Time
}

func newVaultAuth(address, roleID, secretID string, hc *http.Client) *vaultAuth {
	return &vaultAuth{
		address:  strings.TrimRight(address, "/"),
		roleID:   roleID,
		secretID: secretID,
		http:     hc,
	}
}

// Token returns the current Vault token in a thread-safe manner.
func (a *vaultAuth) Token() string {
	a.mu.RLock()
	defer a.mu.RUnlock()
	return a.tok
}

type vaultAuthResponse struct {
	Auth struct {
		ClientToken   string `json:"client_token"`
		LeaseDuration int    `json:"lease_duration"`
		Renewable     bool   `json:"renewable"`
	} `json:"auth"`
}

// login performs an AppRole login and stores the resulting token.
func (a *vaultAuth) login(ctx context.Context) error {
	body, _ := json.Marshal(map[string]string{
		"role_id":   a.roleID,
		"secret_id": a.secretID,
	})
	url := fmt.Sprintf("%s/v1/auth/approle/login", a.address)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewBuffer(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := a.http.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		data, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("vault approle login status %d: %s", resp.StatusCode, string(data))
	}
	var out vaultAuthResponse
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return err
	}
	if out.Auth.ClientToken == "" {
		return fmt.Errorf("vault approle login: empty client_token")
	}
	a.mu.Lock()
	a.tok = out.Auth.ClientToken
	a.renewable = out.Auth.Renewable
	a.ttl = time.Duration(out.Auth.LeaseDuration) * time.Second
	a.expiresAt = time.Now().Add(a.ttl)
	a.mu.Unlock()
	return nil
}

// renewSelf extends the lifetime of the current token.
func (a *vaultAuth) renewSelf(ctx context.Context) error {
	url := fmt.Sprintf("%s/v1/auth/token/renew-self", a.address)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, nil)
	if err != nil {
		return err
	}
	req.Header.Set("X-Vault-Token", a.Token())
	resp, err := a.http.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		data, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("vault renew-self status %d: %s", resp.StatusCode, string(data))
	}
	var out vaultAuthResponse
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return err
	}
	a.mu.Lock()
	if out.Auth.ClientToken != "" {
		a.tok = out.Auth.ClientToken
	}
	a.renewable = out.Auth.Renewable
	a.ttl = time.Duration(out.Auth.LeaseDuration) * time.Second
	a.expiresAt = time.Now().Add(a.ttl)
	a.mu.Unlock()
	return nil
}

// ensureToken makes sure a usable token is available before a Vault call.
// It is invoked on every encrypt/decrypt: the common case is a cheap in-memory
// check, and a network call (renew or re-login) only happens when the token is
// close to expiry or already gone. There is no background goroutine, so an idle
// service performs no Vault activity.
func (a *vaultAuth) ensureToken(ctx context.Context) error {
	a.mu.RLock()
	tok := a.tok
	ttl := a.ttl
	renewable := a.renewable
	expiresAt := a.expiresAt
	a.mu.RUnlock()

	if tok == "" {
		return a.login(ctx)
	}
	// A non-positive TTL means a non-expiring token; nothing to do.
	if ttl <= 0 {
		return nil
	}
	remaining := time.Until(expiresAt)
	if remaining <= 0 {
		return a.login(ctx)
	}
	// Refresh proactively once less than a quarter of the TTL remains.
	if remaining < ttl/4 {
		if renewable {
			if err := a.renewSelf(ctx); err != nil {
				return a.login(ctx)
			}
			return nil
		}
		return a.login(ctx)
	}
	return nil
}

File: conformance/code/individual/src/main/java/com/digit/individual/client/VaultAuth.java
package com.digit.individual.client;

import tools.jackson.databind.JsonNode;
import tools.jackson.databind.json.JsonMapper;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Manages a Vault token obtained via AppRole login and refreshes it on demand
 * (see {@link #ensureToken()}) — there is no background thread. The service only
 * ever performs encrypt/decrypt; all key lifecycle (rotation, config) is managed
 * by Vault itself. Mirrors Go internal/clients/vault_auth.go (Option B renewal).
 */
public class VaultAuth {

    private final String address;
    private final String roleId;
    private final String secretId;
    private final HttpClient httpClient;
    private final JsonMapper mapper;

    private final Object lock = new Object();
    private volatile String token;
    private long ttlSeconds;
    private boolean renewable;
    private Instant expiresAt = Instant.EPOCH;

    public VaultAuth(String address, String roleId, String secretId, HttpClient httpClient, JsonMapper mapper) {
        this.address = trimRight(address);
        this.roleId = roleId;
        this.secretId = secretId;
        this.httpClient = httpClient;
        this.mapper = mapper;
    }

    /** Returns the current Vault token. */
    public String token() {
        return token;
    }

    /** Performs an AppRole login and stores the resulting token. */
    public void login() {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("role_id", roleId);
        body.put("secret_id", secretId);
        JsonNode auth = call(address + "/v1/auth/approle/login", body, null).get("auth");
        applyAuth(auth, true);
    }

    /** Extends the lifetime of the current token via renew-self. */
    private void renewSelf() {
        JsonNode auth = call(address + "/v1/auth/token/renew-self", new LinkedHashMap<>(), token).get("auth");
        applyAuth(auth, false);
    }

    private void applyAuth(JsonNode auth, boolean requireToken) {
        if (auth == null) {
            throw new RuntimeException("vault auth: missing auth block in response");
        }
        JsonNode ctNode = auth.get("client_token");
        String clientToken = ctNode == null ? null : ctNode.asString();
        if (requireToken && (clientToken == null || clientToken.isEmpty())) {
            throw new RuntimeException("vault approle login: empty client_token");
        }
        synchronized (lock) {
            if (clientToken != null && !clientToken.isEmpty()) {
                this.token = clientToken;
            }
            JsonNode renew = auth.get("renewable");
            this.renewable = renew != null && Boolean.parseBoolean(renew.asString());
            JsonNode lease = auth.get("lease_duration");
            long ttl = 0;
            if (lease != null) {
                try {
                    ttl = Long.parseLong(lease.asString());
                } catch (NumberFormatException ignored) {
                    ttl = 0;
                }
            }
            this.ttlSeconds = ttl;
            this.expiresAt = Instant.now().plusSeconds(ttl);
        }
    }

    /**
     * Ensures a usable token is available before a Vault call. The common case is a
     * cheap in-memory check; a network call (renew or re-login) only happens when the
     * token is close to expiry or already gone.
     */
    public void ensureToken() {
        String tok;
        long ttl;
        boolean canRenew;
        Instant exp;
        synchronized (lock) {
            tok = this.token;
            ttl = this.ttlSeconds;
            canRenew = this.renewable;
            exp = this.expiresAt;
        }

        if (tok == null || tok.isEmpty()) {
            login();
            return;
        }
        // A non-positive TTL means a non-expiring token; nothing to do.
        if (ttl <= 0) {
            return;
        }
        long remaining = Instant.now().until(exp, ChronoUnit.SECONDS);
        if (remaining <= 0) {
            login();
            return;
        }
        // Refresh proactively once less than a quarter of the TTL remains.
        if (remaining < ttl / 4) {
            if (canRenew) {
                try {
                    renewSelf();
                } catch (RuntimeException e) {
                    login();
                }
            } else {
                login();
            }
        }
    }

    private JsonNode call(String url, Map<String, Object> body, String vaultToken) {
        String json;
        try {
            json = mapper.writeValueAsString(body);
        } catch (Exception e) {
            throw new RuntimeException("failed to marshal auth request: " + e.getMessage());
        }

        HttpResponse<String> resp;
        try {
            HttpRequest.Builder builder = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .timeout(Duration.ofSeconds(10))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(json));
            if (vaultToken != null) {
                builder.header("X-Vault-Token", vaultToken);
            }
            resp = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        } catch (Exception e) {
            throw new RuntimeException("vault auth request failed: " + e.getMessage());
        }

        if (resp.statusCode() != 200) {
            throw new RuntimeException("vault auth status " + resp.statusCode() + ": " + resp.body());
        }
        try {
            return mapper.readTree(resp.body());
        } catch (Exception e) {
            throw new RuntimeException("failed to decode auth response: " + e.getMessage());
        }
    }

    private static String trimRight(String s) {
        while (s != null && s.endsWith("/")) s = s.substring(0, s.length() - 1);
        return s;
    }
}
